### PR TITLE
boot script

### PR DESCRIPTION
**Pull request by frankiebee**
This adds a sort of "boot script" to the `dev` dir. It will jump start and deploy and fund faucets
it also fixs a bug with registering where we were getting "stale" events for a previous registration attempt found while writing this script includes this doc add in the `dev/README.md`:

## sdk "boot script"

For convince at [./deploy-faucets.mjs](./deploy-facuets.mjs ) is a script we use to "deploy" and fund the faucets for our entropy network you can refer to this section and the script itself to deploy your own faucets! Here is it running with a dev environment setup from the root of this project:

```bash
dev/bin/spin-up.sh four-nodes
```
I would take a deep breath here to allow for the 4 nodes to connect before running the next command bellow. On that note the deploy faucet script takes 3 arguments as you can see right now this script assumes you know what your doing if you are running it. So not a lot of bells and whitelist here.

the first argument is the `endpoint` to target, the second is the `fundingSeed` and the third `faucetLookUpSeed` 

```bash
node dev/deploy-facuets.mjs ws://127.0.0.1:9944 0x786ad0e2df456fe43dd1f91ebca22e235bc162e0bb8d53c633e8c85b2af68b7a 0x20423b5ff4984bcb8922483c98afb7eaa056c40fc431f8a314211e3d94a4222f

```
this example above should run in a dev enviroment the funding seed is eve and the report looks something like this and will log to your console:

```
{
  'jump start status at start': 'Ready',
  'using faucet program pointer': '0x3a1d45fecdee990925286ccce71f78693ff2bb27eae62adf8cfb7d3d61e142aa',
  'faucet program pointer from deployment': '0x3a1d45fecdee990925286ccce71f78693ff2bb27eae62adf8cfb7d3d61e142aa',
  'faucet config': {
    max_transfer_amount: 20000000000,
    genesis_hash: 'a4b29c6895ae775fd291377fde31882f66244eaecbdc81e017ebb64d13b27b72'
  },
  'initial balance for funding account': 99999880954644628n,
  'initial funding faucet amount': 24999970238661157n,
  'modifiableKeys on chain': [
    '0x03cd98af667e48b4912c66576f5cdf18aee3764c7aad1c40b9c61d5ac012acf1f6',
    '0x0228aba3e529d3b78b0cd7454a5f694eb09a648ec488b0b4b8ce7cd9abedd96c28',
    '0x03b7c87542f57895d37f1a37a3460e27ec74de7216e6ed48609ffd2fac976ea94a'
  ],
  'faucet look up address': '5EqZMUYz7jjaG2baQWJRzUzM7YhBP4E8TAj6GgqDqWdXriTn',
  faucets: [
    {
      vk: '0x03cd98af667e48b4912c66576f5cdf18aee3764c7aad1c40b9c61d5ac012acf1f6',
      address: '5Gm6JA2ikMK1FfNn3MRmUPLWfi4p6eBzcFtKE1dnJvQpJgpg',
      balance: '24,999,970,238,661,157'
    },
    {
      vk: '0x0228aba3e529d3b78b0cd7454a5f694eb09a648ec488b0b4b8ce7cd9abedd96c28',
      address: '5FqourMb6c3z4rogDnaBb36Ku53ndy3eCiSdjRg2TUAztJ3X',
      balance: '24,999,970,238,661,157'
    },
    {
      vk: '0x03b7c87542f57895d37f1a37a3460e27ec74de7216e6ed48609ffd2fac976ea94a',
      address: '5CUQuSMxTzx74Zb8gsDobhaYSte9vt9a3Db5X6oio4X1pSX1',
      balance: '24,999,970,238,661,157'
    }
  ],
  endpoint: 'ws://127.0.0.1:9944'
}
```

### CODE REVIEW

File: dev/deploy-facuets.mjs [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+const endpoint = process.argv[2]
+const fundingSeed = process.argv[3]
+const faucetLookUpSeed = process.argv[4]
+// change this number to deploy more faucets
+const faucetCount = 3
+// change me if you change the schemas!
+const pointer = '0x3a1d45fecdee990925286ccce71f78693ff2bb27eae62adf8cfb7d3d61e142aa'
+if (!endpoint) throw new Error('please provide arguments for endpoint, fundingSeed, faucetLookUpSeed')
+if (!fundingSeed) throw new Error('please provide arguments for fundingSeed, faucetLookUpSeed')
+if (!faucetLookUpSeed) throw new Error('please provide arguments for faucetLookUpSeed')
```
Comment (rh0delta) [inline dev/deploy-facuets.mjs:15]:
```suggestion
if (!endpoint || !fundingSeed || !faucetLookUpSeed) throw new Error('endpoint, fundingSeed, and faucetLookupSeed are required arguments')
```

File: dev/deploy-facuets.mjs [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+checkSeed(faucetLookUpSeed)
+checkSeed(fundingSeed)
+
+const faucetAddresses = []
+// actually run the function!
+deployAndFundFaucet().then(() => process.exit(0)).catch((e) => {
+  console.warn(report)
+  console.error(e)
+  process.exit(1)
+})
```
Comment (rh0delta) [inline dev/deploy-facuets.mjs:41]: do we only want to log the report on a failed run of the boot script? might be good to log it for successful attempts too no?

**Review by rh0delta — CHANGES_REQUESTED** [review]

File: dev/deploy-facuets.mjs (hunk 2 above)
Comment (frankiebee) [reply dev/deploy-facuets.mjs:41]: it's logged for both

File: dev/README.md [hunk 3]
```diff
@@ -152,3 +152,58 @@ yarn
 yarn version --patch # patch|minor|major
 npm publish
 ```
+
+## sdk "boot script"
```
Comment (mixmix) [inline dev/README.md:156]: Could open this with context like:

---

When a new network of nodes is turned on, we need some actions to be taken before this new Entropy network is ready to use for tests.

1. Give the nodes a moment to establish connections
2. run "jump start", which establishes who the initial signing nodes are (the network automatically "reshares" these roles periodically after jump start)
3. set up faucet program
    - a) deploy faucet program to network
    - b) install the faucet program for some faucet accounts
    - c) fund the faucet accounts

File: dev/README.md [hunk 4]
```diff
@@ -152,3 +152,58 @@ yarn
...
+
```
Comment (mixmix) [inline dev/README.md:157]:
```suggestion


When a new network of nodes is turned on, we need some actions to be taken before this new Entropy network is ready to use for tests.

1. Give the nodes a moment to establish connections
2. run "jump start", which establishes who the initial signing nodes are (the network automatically "reshares" these roles periodically after jump start)
3. set up faucet program
    - a) deploy faucet program to network
    - b) install the faucet program for some faucet accounts
    - c) fund the faucet accounts

```

File: dev/README.md [hunk 5]
```diff
@@ -152,3 +152,58 @@ yarn
...
+```
+{
+  'jump start status at start': 'Ready',
+  'using faucet program pointer': '0x3a1d45fecdee990925286ccce71f78693ff2bb27eae62adf8cfb7d3d61e142aa',
+  'faucet program pointer from deployment': '0x3a1d45fecdee990925286ccce71f78693ff2bb27eae62adf8cfb7d3d61e142aa',
+  'faucet config': {
+    max_transfer_amount: 20000000000,
+    genesis_hash: 'a4b29c6895ae775fd291377fde31882f66244eaecbdc81e017ebb64d13b27b72'
+  },
+  'initial balance for funding account': 99999880954644628n,
```
[the remark this answers is not in the document]
Comment (mixmix) [reply dev/README.md:182]: or just an inline comment here saying what that unit is. This doesn't need to be super pretty, it's diagnostic

File: dev/deploy-faucets.mjs [hunk 6]
```diff
@@ -0,0 +1,179 @@
+import { readFileSync } from 'fs'
+import { blake2AsHex, encodeAddress } from "@polkadot/util-crypto";
+import Keyring from '@entropyxyz/sdk/keys'
+import Entropy, { wasmGlobalsReady } from '@entropyxyz/sdk'
+import { jumpStartNetwork, createTimeLogProxy } from '@entropyxyz/sdk/testing'
+const endpoint = process.argv[2]
+const fundingSeed = process.argv[3]
+const faucetLookUpSeed = process.argv[4]
```
Comment (mixmix) [inline dev/deploy-faucets.mjs:8]:
```suggestion
const [_, __, endpoint, fundingSeed, faucetLookupSeed] = process.argv
```

File: dev/deploy-faucets.mjs [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+// change this number to deploy more faucets
+const faucetCount = 3
+// change me if you change the schemas!
+const pointer = '0x3a1d45fecdee990925286ccce71f78693ff2bb27eae62adf8cfb7d3d61e142aa'
+if (!endpoint) throw new Error('please provide arguments for endpoint, fundingSeed, faucetLookUpSeed')
+if (!fundingSeed) throw new Error('please provide arguments for fundingSeed, faucetLookUpSeed')
+if (!faucetLookUpSeed) throw new Error('please provide arguments for faucetLookUpSeed')
```
Comment (mixmix) [inline dev/deploy-faucets.mjs:15]: move these into `checkEndpoint` functions etc

File: dev/deploy-faucets.mjs [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+  const faucetEntropy = new Entropy({
+    endpoint,
+    keyring: faucetRing
+  })
+  await faucetEntropy.ready
+  await moneyBags.ready
+  const jumpStartStatus = (await moneyBags.substrate.query.stakingExtension.jumpStartProgress()).toHuman().jumpStartStatus
+  report['jump start status at start'] = jumpStartStatus
+  // deploy faucet program to chain if not already up
+  if (jumpStartStatus === 'Ready') await jumpStartNetwork(moneyBags, true)
```
Comment (mixmix) [inline dev/deploy-faucets.mjs:69]: Break the steps up a bit more with space + comments?

```suggestion

  // Jump Start
  const jumpStartStatus = (await moneyBags.substrate.query.stakingExtension.jumpStartProgress()).toHuman().jumpStartStatus
  report['jump start status at start'] = jumpStartStatus
  if (jumpStartStatus === 'Ready') await jumpStartNetwork(moneyBags, true)
  
```

File: dev/deploy-faucets.mjs [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+        amount: { type: "number" },
+        string_account_id: { type: "string" },
+        spec_version: { type: "number" },
+        transaction_version: { type: "number" },
+      }
+    }
+    report['faucet program pointer from deployment'] = await moneyBags.programs.dev.deploy(faucetProgram, configurationSchema, auxDataSchema)
+  }
+  // transfer funds to faucet account "enough" to register (5 whole tokens)
+  await sendMoney(moneyBags, faucetRing.accounts.registration.address, BigInt(5 * 10 ** 10))
```
Comment (mixmix) [inline dev/deploy-faucets.mjs:95]: Do we have the `1e10` / `10**10` constant defined anywhere in this Repo
I'd like to see at least at top of file

```
const BITS_PER_TOKEN = 1e10
```

```suggestion
  await sendMoney(moneyBags, faucetRing.accounts.registration.address, BigInt(5 * BITS_PER_TOKEN))
```

File: dev/deploy-faucets.mjs [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+import Keyring from '@entropyxyz/sdk/keys'
+import Entropy, { wasmGlobalsReady } from '@entropyxyz/sdk'
+import { jumpStartNetwork, createTimeLogProxy } from '@entropyxyz/sdk/testing'
+const endpoint = process.argv[2]
+const fundingSeed = process.argv[3]
+const faucetLookUpSeed = process.argv[4]
+// change this number to deploy more faucets
+const faucetCount = 3
+// change me if you change the schemas!
+const pointer = '0x3a1d45fecdee990925286ccce71f78693ff2bb27eae62adf8cfb7d3d61e142aa'
```
Comment (mixmix) [inline dev/deploy-faucets.mjs:12]: style: I like these things as CONST so when I see them I know to look at top of file
```suggestion
const FAUCET_COUNT = 3
// change me if you change the schemas!
const POINTER = '0x3a1d45fecdee990925286ccce71f78693ff2bb27eae62adf8cfb7d3d61e142aa'
```

File: dev/deploy-faucets.mjs (hunk 9 above)
Comment (mixmix) [reply dev/deploy-faucets.mjs:95]: This number should be related to the `FAUCET_COUNT` ??

File: dev/deploy-faucets.mjs [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+    faucets.push({
+      'verification key': vk,
+      address: faucetAddress,
+      balance: balance.free.toLocaleString()
+    })
+    report['faucets'] = faucets
+    --faucetCountDown
+  }
+  const modifiableKeys = await moneyBags.substrate.query.registry.modifiableKeys(faucetRing.accounts.registration.address)
+  report['modifiableKeys on chain'] = modifiableKeys.toHuman()
```
Comment (mixmix) [inline dev/deploy-faucets.mjs:134]:
```suggestion
  const verifiyingKeys = await moneyBags.substrate.query.registry.modifiableKeys(faucetRing.accounts.registration.address)
  report['verifyingKeys on chain'] = verifyingKeys.toHuman()
  // These should be the same as faucets.map(faucet => faucet['verification key'])
```

File: dev/deploy-faucets.mjs (hunk 6 above)
Comment (rh0delta) [reply dev/deploy-faucets.mjs:9]: not sure if this is a necessary improvement at this moment, we can clean up the code after the initial release, and just push out a minor patch

File: dev/deploy-faucets.mjs (hunk 10 above)
Comment (rh0delta) [reply dev/deploy-faucets.mjs:13]: i don't think style updates are needed right now to get this out the door, we can polish the boot script after the release sicne we don't really have that many devs that are going to be using the script at the moment

File: dev/fun-bucket.mjs [hunk 12]
```diff
@@ -0,0 +1,20 @@
+const monkeys = ['🙉 - pandemonium', '🙈 - chaos', '🙊 - anarchy', '🐵 - entropy']
```
Comment (rh0delta) [inline dev/fun-bucket.mjs:1]: great addition

File: src/programs/dev.ts [hunk 13]
```diff
@@ -78,7 +78,7 @@ export default class ProgramDev extends ExtrinsicBaseClass {
     const responseOption = await this.substrate.query.programs.programs(pointer)
 
     const programInfo = responseOption.toJSON()
-
+    if (programInfo === null) return null
```
[the remark this answers is not in the document]
Comment (rh0delta) [reply src/programs/dev.ts:81]: program info is null if the program doesn't exist, which is still a valid return type so i don;t believe it constitutes an error

File: src/registration/index.ts [hunk 14]
```diff
@@ -108,49 +108,13 @@ export default class RegistrationManager extends ExtrinsicBaseClass {
...
     })
-    const dataFromEvents = this.#getVerifiyingKeyFromRegisterEvent(
-      this.signer.pair.address
-    )
 
-    await registrationTxResult
-    const verifyingKey = await dataFromEvents
+    const result = await registrationTxResult
+    // @ts-ignore: not sure where the void is coming from
+    const verifyingKey = result.toHuman().event.data[1]
```
[the remark this answers is not in the document]
Comment (rh0delta) [reply src/registration/index.ts:114]: we don't use the address here, not sure if this is a valid updated for now

**Review by rh0delta — DISMISSED** [review]

File: dev/deploy-faucets.mjs (hunk 10 above)
Comment (mixmix) [reply dev/deploy-faucets.mjs:13]: falls in category "important but not urgent" - if we align our code styles a little we get faster parsing + code reviews.

Comment (frankiebee) [reply dev/deploy-faucets.mjs:13]: yes will do this my side so i can make sure it's every where

File: dev/deploy-faucets.mjs (hunk 9 above)
Comment (frankiebee) [reply dev/deploy-faucets.mjs]: done

File: dev/faucet_program.wasm
[the remark this answers is not in the document]
Comment (frankiebee) [reply dev/faucet_program.wasm:1]: i'm kicking down the road this one

File: src/index.ts [hunk 15]
```diff
@@ -125,14 +125,14 @@ export default class Entropy {
...
-   * @param {SS58Address} [params.programDeployer] - The account authorized to modify programs on behalf of the user.
+   * @param {SS58Address} [params.programModAddress] - The account authorized to modify programs on behalf of the user.
    * @param {ProgramInstance[]} [params.programData] - Optional initial programs associated with the user.
    * @returns {Promise<HexString>} A promise that resolves to the verifying key for the new account when the registration is complete.
    * @throws {Error} If the address is already registered or if there's a problem during registration.
    */
   async register (params?: RegistrationParams): Promise<HexString> {
     params = params || this.#getRegisterParamsDefault()
-    if (params.programDeployer && !isDeployer(params.programDeployer)) {
+    if (params.programModAddress && !isDeployer(params.programModAddress)) {
```
[the remark this answers is not in the document]
Comment (frankiebee) [reply src/index.ts:135]: i did put a change log entry for this

File: src/programs/dev.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (frankiebee) [reply src/programs/dev.ts:81]: it's null if it's not on chain

**Review by mixmix — APPROVED** [review]
Looks good to me

**Pull request merged**